src/nvim/auevents.lua

```diff
@@ -72,6 +72,7 @@ return {
     'InsertLeavePre',         -- just before leaving Insert mode
     'LspAttach',              -- after an LSP client attaches to a buffer
     'LspDetach',              -- after an LSP client detaches from a buffer
+    'LspSemanticTokens',      -- after an LSP semantic tokens request is processed
```

Comment:
We are already using `LspRequest` as a User autocommand, maybe we should just promote that to a first class autocmd and make the autocommand pattern match on the request type.

So rather than a new `LspSemanticTokens` autocommand, one would use `autocmd LspRequest textDocument/semanticTokens/*`.

Reply:
Can you direct a patterned autocmd to a specific buffer? Documentation (and reality) seems to imply you can’t mix them.

Reply:
Correct, you cannot. The pattern needs to be a filename in order to do that.

However, we now have support for passing arbitrary data to Lua autocommand callbacks. So we could use that to pass the request to the autocommand callback. We already do this for `LspAttach` and `LspDetach`.

Reply:
So you're suggesting adding a "pattern" to the arbitrary data table so consumers of the generic autocmd can distinguish one for semantic tokens? Something like:

```lua
api.nvim_exec_autocmds('LspResponse', {
  buffer = bufnr,
  data = {
    method = 'textDocument/semanticTokens', -- add /full? /full/*? Not sure what this should be exactly assuming range requests are added at some point 
    highlights = buf_state.highlights,
  }
})
```

`LspRequest` itself doesn't work because that's fired when requests are sent; not when responses are received. We could promote both for symmetry.

Reply:
I don't think `LspRequest` is a good match here, it maps 1:1 to lsp requests, but the token event is synthesized.
If we're going to use autocmds I think we'll need a new dedicated event.

An alternative could be to move away from the extensiblity/semantic-token abstraction level, and instead go for a  high level LSP highlighting API/functionality. Without allowing access to tokens and also hiding the LSP semantic token semantics. E.g. something like `vim.lsp.hl_attach(client, buf)` that you can call within a `LspAttach` event and that keeps everything else implementation detail, without any other public API.

I'm not sure if that's better or worse, but it would focus the API more on the common case users are going to use it for, and plugins who really want to use the tokens could still use the lsp client's request functions directly to retrieve them.

> Determine if core should provide basic highlighting support

Yes I think it should.

Reply:
Closing since this comment is no longer relevant to the design.